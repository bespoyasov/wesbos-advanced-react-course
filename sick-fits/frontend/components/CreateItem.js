import React, {Component} from 'react'
import {Mutation} from 'react-apollo'
import gql from 'graphql-tag'
import Router from 'next/router'

import Form from './styles/Form'
import Error from './ErrorMessage'
import format from '../lib/formatMoney'

import {
  imageUploadPreset as preset,
  imageUploadEndpoint as endpoint,
} from '../config'


export const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`

class CreateItem extends Component {
  state = {
    title: 'Some title',
    description: 'Some description',
    image: '',
    largeImage: '',
    price: 1500,
  }

  handleChange = e => {
    const {name, type, value} = e.target
    const val = type === 'number' ? parseFloat(value) : value
    return this.setState(state => ({ [name]: val }))
  }

  uploadFile = async e => {
    const files = e.target.files
    const data = new FormData()
    data.append('file', files[0])
    data.append('upload_preset', preset)

    const response = await fetch(endpoint, { method: 'POST', body: data })
    const file = await response.json()

    return this.setState(state => ({ 
      image: file.secure_url,
      largeImage: file.eager[0].secure_url,
    }))
  }

  render() {
    const {title, price, description, image} = this.state

    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
        {(createItem, {loading, error}) => (
          <Form onSubmit={async e => {
            e.preventDefault() 
            const response = await createItem()
            Router.push({
              pathname: '/item',
              query: {id: response.data.createItem.id}
            })
          }}>
            <Error error={error} />
            
            <fieldset disabled={loading} aria-busy={loading}>
              <label>
                Upload an image
                <input 
                  type='file' 
                  name='file' 
                  placeholder='Upload an image' 
                  onChange={this.uploadFile}
                  required />

                {!!image 
                  && <img 
                      alt='Upload preview'
                      width='200'
                      src={image} />}
              </label>

              <label>
                Title
                <input 
                  type='text' 
                  name='title' 
                  placeholder='Title' 
                  value={title}
                  onChange={this.handleChange}
                  required />
              </label>

              <label>
                Price
                <input 
                  type='number' 
                  name='price' 
                  placeholder='Price' 
                  value={price}
                  onChange={this.handleChange}
                  required />
              </label>

              <label>
                Description
                <textarea 
                  name='description' 
                  placeholder='Description' 
                  value={description}
                  onChange={this.handleChange}
                  required />
              </label>

              <button type='submit'>Submit</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    )
  }
}

export default CreateItem