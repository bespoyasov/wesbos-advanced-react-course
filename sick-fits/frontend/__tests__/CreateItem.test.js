import {mount} from 'enzyme'
import wait from 'waait'
import toJSON from 'enzyme-to-json'
import {MockedProvider} from 'react-apollo/test-utils'
import Router from 'next/router'

import {fakeItem} from '../lib/testUtils'
import CreateItem, {
  CREATE_ITEM_MUTATION as query,
} from '../components/CreateItem'


Router.router = {
  push: jest.fn()
}


const dogImage = 'https://dog.com/dog.jpg'

global.fetch = jest.fn().mockResolvedValue({
  json: () => ({
    secure_url: dogImage,
    eager: [{secure_url: dogImage}],
  })
})


describe('<CreateItem />', () => {
  it('renders and matches snapshot', () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    )

    const form = wrapper.find('form[data-test="create-item-form"]')
    expect(toJSON(form)).toMatchSnapshot()
  })

  it('uploads file when change', async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    )

    const fileInput = wrapper.find('input[type="file"]')
    fileInput.simulate('change', {
      target: {
        files: ['fakeDog.jpg']
      }
    })

    await wait()
    const component = wrapper.find('CreateItem').instance()
    expect(component.state.image).toEqual(dogImage)
    expect(component.state.largeImage).toEqual(dogImage)
    
    expect(global.fetch).toHaveBeenCalled()
    global.fetch.mockReset()
  })

  it('handles state updating', async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    )

    wrapper
      .find('[name="title"]')
      .simulate('change', {target: {value: 'test', name: 'title'}})
    
    wrapper
      .find('[name="price"]')
      .simulate('change', {target: {value: 50000, name: 'price', type: 'number'}})
    
    // etc
    wrapper.update()
    
    expect(wrapper.find('CreateItem').instance().state).toMatchObject({
      title: 'test',
      price: 50000
    })
  })

  it('creates an item when the form is submitted', async () => {
    const item = fakeItem()
    
    const mocks = [{
      request: {
        query, 
        variables: {
          title: item.title,
          description: item.description,
          image: '',
          largeImage: '',
          price: item.price,
        }
      },
      result: {
        data: {
          createItem: {
            ...item,
            __typename: 'item'
          }
        }
      }
    }]

    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <CreateItem />
      </MockedProvider>
    )

    wrapper
      .find('[name="title"]')
      .simulate('change', {target: {value: item.title, name: 'title'}})
    
    wrapper
      .find('[name="price"]')
      .simulate('change', {target: {value: item.price, name: 'price', type: 'number'}})

    wrapper
      .find('[name="description"]')
      .simulate('change', {target: {value: item.description, name: 'description'}})

    wrapper.find('form').simulate('submit')

    await wait(50)
    expect(Router.router.push).toHaveBeenCalledWith({
      pathname: "/item", 
      query: {id: "abc123"}
    })
  })
})