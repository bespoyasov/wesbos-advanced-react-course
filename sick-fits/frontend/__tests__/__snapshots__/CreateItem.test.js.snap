// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateItem /> renders and matches snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 krRxky"
  data-test="create-item-form"
  onSubmit={[Function]}
>
  <DisplayError
    error={Object {}}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label>
      Upload an image
      <input
        name="file"
        onChange={[Function]}
        placeholder="Upload an image"
        required={true}
        type="file"
      />
    </label>
    <label>
      Title
      <input
        name="title"
        onChange={[Function]}
        placeholder="Title"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label>
      Price
      <input
        name="price"
        onChange={[Function]}
        placeholder="Price"
        required={true}
        type="number"
        value={0}
      />
    </label>
    <label>
      Description
      <textarea
        name="description"
        onChange={[Function]}
        placeholder="Description"
        required={true}
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Submit
    </button>
  </fieldset>
</form>
`;
